src/QsCompiler/SyntaxProcessor/CapabilityInference.fs

```diff
@@ -193,48 +196,85 @@ let private scopePatterns scope = scope.Statements |> Seq.collect statementPatte
     | _ -> baseCapability
 
-/// Returns the required capability of the callable.
-let private callableCapability callable =
-    let inOperation =
-        match callable.Kind with
-        | Operation -> true
-        | _ -> false
-    callable.Specializations |> Seq.map (specializationCapability inOperation) |> maxCapability
-
-/// Adds the inferred capability of the callable as an attribute.
-let private addAttribute callable =
-    let capability = callableCapability callable
+/// Returns the required capability of the specialization based on its source code and callable dependencies.
+let rec private specDependentCapability
+        (callables : IImmutableDictionary<_, _>) (graph : CallGraph) visited (spec : QsSpecialization) =
+    let visited = Set.add spec.Parent visited
+    let inOperation = callables.TryGetValue spec.Parent |> tryToOption |> Option.exists isOperation
+    let dependencies =
+        graph.GetDirectDependencies spec
+        |> Seq.map (fun group -> group.Key.CallableName)
+        |> Set.ofSeq
+        |> fun names -> Set.difference names visited
+    dependencies
+    |> Seq.choose (callables.TryGetValue >> tryToOption)
+    |> Seq.map (callableDependentCapability callables graph visited)
+    |> Seq.append (specSourceCapability inOperation spec |> Seq.singleton)
+    |> maxCapability
+
+/// Returns the required capability of the callable based on its capability attribute if one is present. If no attribute
+/// is present and the callable is not defined in a reference, returns the capability based on its source code and
+/// callable dependencies. Otherwise, returns the base capability.
+and private callableDependentCapability callables graph visited (callable : QsCallable) =
+    callable.Attributes
+    |> QsNullable<_>.Choose BuiltIn.GetCapability
+    |> tryMaxCapability
+    |> Option.defaultWith (fun () ->
+        if callable.SourceFile.Value.EndsWith ".qs"
+        then
+            callable.Specializations
+            |> Seq.map (specDependentCapability callables graph visited)
+            |> maxCapability
+        else baseCapability)
+
+/// Adds the given capability to the callable as an attribute.
+let private addAttribute callable capability =
     let arg = capability.ToString () |> AttributeUtils.StringArgument
     let attribute = AttributeUtils.BuildAttribute (BuiltIn.Capability.FullName, arg)
-    { callable with Attributes = callable.Attributes.Add attribute }
+    { callable with QsCallable.Attributes = callable.Attributes.Add attribute }
 
 /// Infers the capability of all callables in the compilation, adding the built-in Capability attribute to each
 /// callable.
-///
-/// TODO:
-/// This infers the capability of each callable individually, without looking at the capabilities of anything it
-/// calls or references. The inferred capability is only a lower bound.
 let InferCapabilities compilation =
```

Comment:
How is this meant to be used once the capability inference feature is completed? For any given Q# project, will this be called for the generated QsCompilation?

Reply:
`InferCapabilities` is called automatically by a rewrite step for Q# libraries. It's the only place that I intend to call it from.

https://github.com/microsoft/qsharp-compiler/blob/370f6bf5052af25b28291905be4c7077bc6e62eb/src/QsCompiler/Compiler/RewriteSteps/CapabilityInference.cs#L34-L38

Reply:
Thanks Sarah! Have you tried compiling the standard library with these changes using the `--perf` option? It would be good to double check that `OverallCompilation.RewriteSteps` does not take much more time after the change.

Reply:
Thanks for the suggestion. Here is a run with capability inference enabled:

```
{
  "OverallCompilation": 58794,
  "OverallCompilation.Build": 38276,
  "OverallCompilation.OutputGeneration": 3332,
  "OverallCompilation.OutputGeneration.BinaryGeneration": 30,
  "OverallCompilation.OutputGeneration.DocumentationGeneration": 722,
  "OverallCompilation.OutputGeneration.SyntaxTreeSerialization": 2580,
  "OverallCompilation.ReferenceLoading": 2297,
  "OverallCompilation.RewriteSteps": 14782,
  "OverallCompilation.SourcesLoading": 73
}
```

and one with capability inference disabled:

```
{
  "OverallCompilation": 50164,
  "OverallCompilation.Build": 38713,
  "OverallCompilation.OutputGeneration": 3505,
  "OverallCompilation.OutputGeneration.BinaryGeneration": 26,
  "OverallCompilation.OutputGeneration.DocumentationGeneration": 733,
  "OverallCompilation.OutputGeneration.SyntaxTreeSerialization": 2745,
  "OverallCompilation.ReferenceLoading": 2008,
  "OverallCompilation.RewriteSteps": 5822,
  "OverallCompilation.SourcesLoading": 79
}
```

So, it does have a significant impact on the rewrite step time. That is expected since this algorithm is not very efficient and is intended to be replaced with a more efficient one later (since the more efficient algorithm I had in mind requires changes to the call graph). But I can try the ideas @ScottCarda-MSFT mentioned for increasing the performance to see if that works without changing the call graph.

Reply:
I updated the algorithm. This one is much faster. Here is a new run with capability inference enabled:

```
{
  "OverallCompilation": 53029,
  "OverallCompilation.Build": 40176,
  "OverallCompilation.OutputGeneration": 3792,
  "OverallCompilation.OutputGeneration.BinaryGeneration": 27,
  "OverallCompilation.OutputGeneration.DocumentationGeneration": 798,
  "OverallCompilation.OutputGeneration.SyntaxTreeSerialization": 2965,
  "OverallCompilation.ReferenceLoading": 2390,
  "OverallCompilation.RewriteSteps": 6535,
  "OverallCompilation.SourcesLoading": 81
}
```

So it's adding about 700ms to the rewrite steps (very roughly - I didn't bother to average over multiple runs).

Reply:
It would be worth trying @ScottCarda-MS idea. Even though the current algorithm is not very efficient, making it a bit better would be great.

Reply:
Yes, I did update the algorithm based on @ScottCarda-MS's suggestion - I think it's asymptotically optimal now.

Reply:
That's great! It looks much better now. Thanks for making the change.